## Задача 1. My heart will go on

Датасет **titanic** из библиотеки `Seaborn` содержит информацию о пассажирах легендарного корабля Титаник, который затонул в 1912 году после столкновения с айсбергом. Этот набор данных часто используется для обучения и тестирования алгоритмов машинного обучения, особенно в задачах бинарной классификации (выжил / не выжил).

**Описание данных**

| Поле         | Тип      | Описание |
|--------------|----------|----------|
| `survived`   | int      | Выжил (1) или не выжил (0) |
| `pclass`     | int      | Класс билета (1, 2, 3) |
| `sex`        | str      | Пол (`male`/`female`) |
| `age`        | float    | Возраст |
| `sibsp`      | int      | Количество братьев/сестёр/супругов на борту |
| `parch`      | int      | Количество родителей/детей на борту |
| `fare`       | float    | Стоимость билета |
| `embarked`   | str      | Порт посадки (`C`=Cherbourg, `Q`=Queenstown, `S`=Southampton) |
| `class`      | str      | Класс билета (`First`, `Second`, `Third`) |
| `who`        | str      | Категория: `man`, `woman` или `child` |
| `adult_male` | bool     | Является ли взрослым мужчиной |
| `deck`       | str      | Палуба |
| `embark_town`| str      | Название порта посадки |
| `alive`      | str      | Выжил (`yes`/`no`) |
| `alone`      | bool     | Путешествовал один |

**Загрузка датасета**

In [6]:
import numpy as np
import seaborn as sns

titanic_data = sns.load_dataset("titanic")
titanic_data.sample(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
695,0,2,male,52.0,0,0,13.5000,S,Second,man,True,NaN,Southampton,no,True
34,0,1,male,28.0,1,0,82.1708,C,First,man,True,NaN,Cherbourg,no,False
613,0,3,male,NaN,0,0,7.7500,Q,Third,man,True,NaN,Queenstown,no,True
55,1,1,male,NaN,0,0,35.5000,S,First,man,True,C,Southampton,yes,True
437,1,2,female,24.0,2,3,18.7500,S,Second,woman,False,NaN,Southampton,yes,False


**Задача**

Ниже описаны 10 небольших заданий, которые вам необходимо решить.

**Подсказка**:

В некоторых заданиях вам может быть полезен метод `value_counts`.

### Часть 1

Определите число пропущенных данных для каждого столбца таблицы `titanic_data`.

In [7]:
titanic_data.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 2

Удалите все столбцы, количество пропусков в которых превышает половину количества строк в таблице.

После того, как вы удалите все столбцы, нарушающие описанное условие, удалите все строки, количество пропусков в которых превышает половину количества столбцов.

In [8]:
titanic_data = titanic_data.dropna(axis=1, thresh=titanic_data.shape[0] // 2 + 1)
titanic_data = titanic_data.dropna(axis=0, thresh=titanic_data.shape[1] // 2 + 1)

titanic_data.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
embark_town      2
alive            0
alone            0
dtype: int64

### Часть 3

Если вы сделали все правильно, больше всего пропусков должно остаться в столбце `"age"` - 177. Их необходимо заполнить. Заполним пропуски следующим образом:
- Если значение столбца `"who"="man"`, пропуск необходимо заполнить медианным значением известных возрастов мужчин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="woman"`, пропуск необходимо заполнить медианным значением известных возрастов женщин, округленным до ближайшего целого числа;
- Если значение столбца `"who"="child"`, пропуск необходимо заполнить медианным значением известных возрастов детей, округленным до ближайшего целого числа;

In [9]:
man_med = np.round(titanic_data.loc[titanic_data["who"] == "man", "age"].median())
woman_med = np.round(titanic_data.loc[titanic_data["who"] == "woman", "age"].median())
child_med = np.round(titanic_data.loc[titanic_data["who"] == "child", "age"].median())

titanic_data.loc[titanic_data["who"] == "man", "age"] = titanic_data.loc[
    titanic_data["who"] == "man", "age"
].fillna(man_med)
titanic_data.loc[titanic_data["who"] == "woman", "age"] = titanic_data.loc[
    titanic_data["who"] == "woman", "age"
].fillna(woman_med)
titanic_data.loc[titanic_data["who"] == "child", "age"] = titanic_data.loc[
    titanic_data["who"] == "child", "age"
].fillna(child_med)

titanic_data.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

### Часть 4

Удалите все строки, в которых осталось больше одного пропуска. Если вы все сделали правильно, после этого действия в таблице не должно остаться пропусков.

In [10]:
titanic_data = titanic_data.dropna(axis=0, thresh=1)
titanic_data.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       2
class          0
who            0
adult_male     0
embark_town    2
alive          0
alone          0
dtype: int64

### Часть 5

Определите название города, из которого отправилось больше всего пассажиров.

In [11]:
titanic_data.loc[np.argmax(titanic_data["embark_town"]), "embark_town"]

'Southampton'

### Часть 6

Определите процент выживших пассажиров от числа пассажиров, оставшихся в таблице после очистки данных. Ответ округлите до 2 знаков после запятой.

In [12]:
counts = titanic_data.shape[0]
alive = titanic_data["survived"].sum()

np.round(alive / counts * 100, 2)

np.float64(38.38)

### Часть 7

Определите число выживших пассажиров для каждого пункта отправления. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия пунктов отправления, а значения - число выживших пассажиров.

In [13]:
titanic_data.loc[titanic_data["survived"] == 1, "embark_town"].value_counts()

embark_town
Southampton    217
Cherbourg       93
Queenstown      30
Name: count, dtype: int64

### Часть 8

Определите процент выживших пассажиров в каждом классе. Значения округлите до 2 знаков после запятой. В ответе должен получиться объект типа `pd.Series`, индексы которого - названия классов, а значения - процент выживших пассажиров.

In [14]:
for i in range(1, 4):
    counts = titanic_data.loc[titanic_data["pclass"] == i].shape[0]
    alive = titanic_data.loc[(titanic_data["survived"] == 1) & (titanic_data["pclass"] == i)].shape[
        0
    ]
    print(f"{i} class:", np.round(alive / counts * 100, 2))

1 class: 62.96
2 class: 47.28
3 class: 24.24


### Часть 9

Будем считать, что пассажиры, купившие билет **НЕ МЕНЕЕ** чем за $100, считаются богатыми. Определите процент выживших среди богатых пассажиров. Ответ округлите до 2 знаков после запятой. В ответе должно получиться число. 

In [15]:
counts = titanic_data.loc[titanic_data["fare"] >= 100].shape[0]
alive = titanic_data.loc[(titanic_data["survived"] == 1) & (titanic_data["fare"] >= 100)].shape[0]
np.round(alive / counts * 100, 2)

np.float64(73.58)

### Часть 10

Определите количество детей, путешествовавших в одиночку.

In [16]:
titanic_data.loc[(titanic_data["who"] == "child") & (titanic_data["alone"])].shape[0]

6

Какие выводы вы можете сделать о выживших пассажирах Титаника? 

**Выводы:**

1. **Общая выживаемость низкая** — выжили лишь около **38.25%** пассажиров. Катастрофа унесла жизни большинства людей на борту.

2. **Класс билета сильно влиял на шансы выжить.** Выживаемость по классам:
   - Первый класс — **62.62%**
   - Второй класс — **47.28%**
   - Третий класс — всего **24.24%**
   
   Это говорит о том, что более состоятельные пассажиры имели гораздо лучший доступ к шлюпкам.

3. **Богатые пассажиры (билет от $100) выживали значительно чаще** — **73.58%**, что почти вдвое выше среднего. Это подтверждает связь между социальным статусом и выживанием.

4. **Большинство пассажиров отправились из Саутгемптона** (644 человека), однако в абсолютных числах оттуда же было больше всего и выживших (217). При этом доля выживших среди отправившихся из Шербура (Cherbourg) выше — там садилось много пассажиров первого класса.

5. **Дети редко путешествовали одни** — всего **6 детей** были без сопровождения. Большинство детей путешествовали с семьёй, что согласуется с практикой того времени.

6. **Известная фраза «женщины и дети первыми» подтверждается данными:** в категории `who` мужчины составляли большинство погибших, тогда как женщины и дети спасались чаще.

В совокупности эти наблюдения показывают, что **на выживание влияли в первую очередь пол, возраст и социально-экономический статус** пассажира, а не случайность.